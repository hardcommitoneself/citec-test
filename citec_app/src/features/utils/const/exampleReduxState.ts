// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.

export const exampleReduxState = {
   errors: '',
   taxHarvesting: {
      arrayPositionSelected: [
         {
            outstanding_balance: 0,
            ticker: '$CASH',
            unrealized_capital_gains: 500,
            taxAlpha: -200,
         },
      ],
      positionsTable: [
         {
            outstanding_balance: 50000,
            ticker: 'AAPL',
            unrealized_capital_gains: -1000,
            taxAlpha: 400,
         },
         {
            outstanding_balance: 0,
            ticker: '$CASH',
            unrealized_capital_gains: 500,
            taxAlpha: -200,
         },
      ],
      taxAlpha: -400,
      unrealizedCapitalGains: -500,
      unrealizedCapitalGainsNegative: -1000,
      unrealizedCapitalGainsSelected: 0,
   },
   portfolio: {
      ticker: ['$CASH', 'AAPL'],
      outstanding_balance: [1000000, 50000],
      unrealized_capital_gains: [0, 11234],
      analysis: {
         fundamentals: {
            total_assets: 2,
            total_positions: 2,
            pe: 24.32,
            pb: 38.96,
            dividend_yield: 0.59,
         },
         risk: {
            sharpe: 0.305683,
            volatility: 0.0147,
            beta: 0.048719,
            alpha: 0.000035,
         },
         returns: {
            returns_1M: -0.0021,
            returns_3M: -0.0037,
            returns_6M: -0.0028,
            CAGR: -0.1058,
         },
         capital_gains: {
            outstanding_balance: 1050000,
            unrealized_cap_losses: 0,
            unrealized_cap_gains: 11234,
            realized_cap_losses: 0,
            realized_cap_gains: 0,
            taxes_cost: 0,
            taxes_rate: 0.4,
         },
         sectors: {
            sector: ['Cash', 'Information Technology'],
            percentage: [0.952381, 0.047619],
            percentage_benchmark: [0.3333333333333333, 0.6666666666666666],
         },
         factors: {
            factor: ['style', 'size'],
            percentage: [0.502991, 0.52381],
         },
         segment_matrix: {
            style: ['Val', 'Val'],
            size: ['Large', 'Mid'],
            weight: [0, 0],
         },
         top_holdings: {
            ticker: ['$CASH', 'AAPL'],
            weight: [0.952381, 0.047619],
            outstanding_balance: [1000000, 50000],
            unrealized_capital_gains: [0, 11234],
         },
         concentration: {
            ticker: ['$CASH', 'AAPL'],
            concentration: [95.238095, 100],
         },
         var: {
            quantile: [2, 4],
            value: [0.000072, 0.001334],
         },
         risk_returns: {
            ticker: ['portfolio', 'benchmark'],
            returns: [0.004722, 0.119508],
            volatility: [0.008561, 0.205196],
            sharpe: [0.551568, 0.582407],
         },
      },
   },
   universe: {
      factors_filter: [
         {
            factor: 'style',
            ranges: [0, 100],
         },
         {
            factor: 'size',
            ranges: [0, 100],
         },
         {
            factor: 'yield',
            ranges: [0, 100],
         },
      ],
      sectors_filter: [
         'Industrials',
         'Financials',
         'Health Care',
         'Consumer Staples',
         'Communication Services',
         'Information Technology',
         'Consumer Discretionary',
         'Materials',
         'Cash',
         'Real Estate',
         'Energy',
         'Utilities',
      ],
      to_exclude: [],
      to_include: [],
      universe_base: 'SPDR S&P 500',
   },
   rebalance: {
      ticker: ['$CASH', 'AAPL'],
      hold: [0, 1],
      objective: {
         name: 'Minimize tracking error',
         key: 'min_tracking_error',
      },
      benchmark: 'SPY',
      min_weight: 0.000001,
      number_stocks: [10, 200],
   },
   portfolio_optimized: {
      portfolio_id: null,
      portfolio: 'optimized_portfolio',
      benchmark: 'SPY',
      creation_date: null,
      ticker: ['AAPL', 'MSFT'],
      outstanding_balance: [72506.52, 63477.37],
      unrealized_capital_gains: [0, 0],
      optimization_status: 'Optimization terminated successfully',
   },
   trades: [
      {
         ticker: 'MRK',
         operation: 'SELL',
         volume: 275161.76,
         realized_capital_gains: 0,
         unrealized_capital_gains: 0,
         stocks_number: 3053,
         expected_volume: 275193,
         delta_volume: 31,
         price: 90.1388,
         columnName: {
            url_logo:
               'https://static.finnhub.io/logo/2eedcaa6-80ec-11ea-853f-00000000092a.png',
            name: 'Merck & Co Inc',
         },
         gain_loss: [0, 0],
      },
      {
         ticker: 'BMY',
         operation: 'SELL',
         volume: 152166.26,
         realized_capital_gains: 0,
         unrealized_capital_gains: 0,
         stocks_number: 2072,
         expected_volume: 152146,
         delta_volume: -19,
         price: 73.43,
         columnName: {
            url_logo:
               'https://static.finnhub.io/logo/8b628318-80eb-11ea-93a2-00000000092a.png',
            name: 'Bristol-Myers Squibb Co',
         },
         gain_loss: [0, 0],
      },
      {
         ticker: 'WMT',
         operation: 'SELL',
         volume: 83898.27,
         realized_capital_gains: 0,
         unrealized_capital_gains: 0,
         stocks_number: 679,
         expected_volume: 83845,
         delta_volume: -53,
         price: 123.4832,
         columnName: {
            url_logo:
               'https://static.finnhub.io/logo/52057505dd05a752d364065fcddf7772a700b31c47c39e9dcb41e393f6693d52.png',
            name: 'Walmart Inc',
         },
         gain_loss: [0, 0],
      },
   ],
   workflow: {
      type: 'load',
   },
   static_data: {
      universe: {
         universes: [
            {
               name: 'INVESCO QQQ TRUST SERIES 1',
               description: 'INVESCO QQQ TRUST SERIES 1',
            },
            {
               name: 'ISHARES MSCI USA ESG SELECT',
               description: 'ISHARES MSCI USA ESG SELECT',
            },
         ],
         sectors: [
            'Cash',
            'Communication Services',
            'Consumer Discretionary',
            'Consumer Staples',
         ],
         factors: ['style', 'size', 'yield', 'quality'],
      },
      rebalance: {
         objectives: [
            {
               key: 'min_tracking_error',
               name: 'Minimize tracking error',
            },
            {
               key: 'min_risk_parity',
               name: 'Minimize risk parity',
            },
            { key: 'min_volatility', name: 'Minimize volatility' },
            {
               key: 'max_diversification_ratio',
               name: 'Maximize diversification ratio',
            },
         ],
         benchmarks: [
            {
               key: 'sp_500',
               description: 'S&P 500 universe',
               ticker: 'SPY',
            },
            {
               key: 'nasdaq_100',
               description: 'NASDAQ 100',
               ticker: 'QQQ',
            },
         ],
      },
   },
   brokers: {
      brokerConnected: '',
      balance: 0,
      accountType: [],
      a1c_t: '',
   },
   ideas: [
      {
         portfolio_id: 16,
         portfolio: 'my_first_portfolio_sadasd',
         benchmark: 'SPY',
         creation_date: '2022-04-06 16:10:19',
         ticker: ['MSFT', 'AAPL'],
         outstanding_balance: [200000, 100000],
         unrealized_capital_gains: [0, 35700],
         optimization_status: null,
      },
      {
         portfolio_id: 18,
         portfolio: 'testing',
         benchmark: 'SPY',
         creation_date: '2022-04-06 16:10:19',
         ticker: ['MSFT', 'AAPL'],
         outstanding_balance: [200000, 100000],
         unrealized_capital_gains: [0, 35700],
         optimization_status: null,
      },
      {
         portfolio_id: 19,
         portfolio: 'Jueves',
         benchmark: 'SPY',
         creation_date: '2022-06-23 10:38:00',
         ticker: [
            'CMCSA',
            'REGN',
            'VTR',
            'RTX',
            'NEE',
            'AAPL',
            'MMM',
            'TMO',
            'BLK',
            'MSFT',
            'QCOM',
            'NVDA',
            'CSX',
            'ADBE',
            'XOM',
            'AVGO',
            'BAC',
            'TSLA',
            'SHW',
            'TXN',
            'NFLX',
            'GOOGL',
            'C',
            'MMC',
            'BKNG',
            'COST',
            'NKE',
            'DIS',
            'EL',
            'CSCO',
            'BRK.B',
            'META',
            'AMZN',
            'DG',
            'CCI',
            'ORCL',
            'USB',
            'TGT',
            'PG',
            'ICE',
            'GOOG',
            'AMD',
            'MCD',
            'CVS',
            'HD',
            'CME',
            'MU',
            'UNH',
            'ATVI',
            'V',
            'EOG',
            'GS',
            'HUM',
            'COP',
         ],
         outstanding_balance: [
            80352.37, 74072.31, 59343.2, 54736.44, 41924.05, 41510.68, 35441.43,
            27349.74, 26308.48, 26136.99, 25646.73, 25479.44, 24726.53, 24075.27,
            23856.18, 23763.38, 23735.12, 23092.02, 22995.95, 21888.81, 21775.5,
            21090.31, 16734.94, 16192.87, 15720.33, 14020.04, 14018.61, 13225.5,
            12996.74, 11693.52, 11315.64, 10798.53, 10583.72, 10309.54, 10245.54,
            9611.94, 8747.21, 8723.18, 6339.86, 5866.02, 5426.94, 4912.02, 4857.23,
            3915.03, 3395.13, 3370.41, 3322.34, 3184.94, 3001.78, 1983.75, 1858.2,
            1789.12, 1309.98, 1228.49,
         ],
         unrealized_capital_gains: [
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0,
         ],
         optimization_status: null,
      },
      {
         portfolio_id: 20,
         portfolio: 'Min volatility',
         benchmark: 'SPY',
         creation_date: '2022-06-23 11:25:41',
         ticker: [
            'MRK',
            'BMY',
            'WMT',
            'CMCSA',
            'REGN',
            'VTR',
            'RTX',
            'NEE',
            'MMM',
            'CSX',
            'XOM',
            'SHW',
            'TXN',
            'C',
            'BRK.B',
            'BA',
            'BSX',
            'CB',
            'CI',
            'CL',
            'CME',
            'COP',
            'COST',
            'CRM',
            'ZTS',
            'ADBE',
            'AMZN',
            'AVGO',
         ],
         outstanding_balance: [
            275161.76, 152166.26, 83898.27, 80352.37, 74072.31, 59343.2, 54736.44,
            41924.05, 35441.43, 24726.53, 23856.18, 22995.95, 21888.81, 16734.94,
            11315.64, 2116.49, 2116.49, 2116.49, 2116.49, 2116.49, 2116.49, 2116.49,
            2116.49, 2116.49, 2116.49, 119.48, 96.79, 4.71,
         ],
         unrealized_capital_gains: [
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0,
         ],
         optimization_status: null,
      },
      {
         portfolio_id: 21,
         portfolio: 'min2',
         benchmark: 'SPY',
         creation_date: '2022-06-23 11:27:32',
         ticker: [
            'MRK',
            'BMY',
            'WMT',
            'CMCSA',
            'REGN',
            'VTR',
            'RTX',
            'NEE',
            'MMM',
            'CSX',
            'XOM',
            'SHW',
            'TXN',
            'C',
            'BRK.B',
            'BA',
            'BSX',
            'CB',
            'CI',
            'CL',
            'CME',
            'COP',
            'COST',
            'CRM',
            'ZTS',
            'ADBE',
            'AMZN',
            'AVGO',
         ],
         outstanding_balance: [
            275161.76, 152166.26, 83898.27, 80352.37, 74072.31, 59343.2, 54736.44,
            41924.05, 35441.43, 24726.53, 23856.18, 22995.95, 21888.81, 16734.94,
            11315.64, 2116.49, 2116.49, 2116.49, 2116.49, 2116.49, 2116.49, 2116.49,
            2116.49, 2116.49, 2116.49, 119.48, 96.79, 4.71,
         ],
         unrealized_capital_gains: [
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0,
         ],
         optimization_status: null,
      },
   ],
}
