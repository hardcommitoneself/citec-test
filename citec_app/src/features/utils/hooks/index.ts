// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.

export * from './use-app-navigate'
export * from './use-app-redux-loader'
export * from './use-app-subscriptions'
export * from './use-debounce'
export * from './use-delete-idea'
export * from './use-fetch-api'
export * from './use-form'
export * from './use-get-auth0-token'
export * from './use-get-custom-universe'
export * from './use-get-optimization'
export * from './use-get-summary'
export * from './use-handler-tax-harvesting'
export * from './use-select-universe'
export * from './use-validate-criteria-selected'
export * from './use-validate-number-stocks'
export * from './use-validate-portfolio-loaded'
export * from './use-validate-registered-user'
export * from './use-validate-sold-position'
export * from './use-validate-ticker'
export * from './use-validate-tickers'
