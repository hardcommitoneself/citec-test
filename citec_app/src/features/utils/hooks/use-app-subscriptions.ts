// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import axios from "axios"
import Swal from "sweetalert2"

export const useAppSubscriptions = (user: any, pathname: string) => {
   const trial_period = process.env.REACT_APP_TRIAL_PERIOD as any

   if (user &&
      process.env.REACT_APP_ACTIVATE_SUBSCRIPTIONS == 'true' &&
      pathname != '/subscriptions') {

      const hasExpirationDate = !!user["https://citecsolutions.com/timestamp_expiration_datess"]
      if (hasExpirationDate) {
         checkIfExpirationDateIsLessToday(user["https://citecsolutions.com/timestamp_expiration_date"])
      } else {
         addTrialPeriod(trial_period)
      }
   }
}


const checkIfExpirationDateIsLessToday = (expiration_date: number) => {
   const timestamp_current = Math.floor(Date.now() / 1000)

   if (expiration_date < timestamp_current) {
      Swal.fire({
         title: 'Your subscription has expired!',
         html: 'The last day you had access was '
            + new Date(expiration_date * 1000).toLocaleDateString("en-US"),
         timer: 6000,
         timerProgressBar: true,
         didOpen: () => {
            Swal.showLoading()
         },
         willClose: () => {
            window.location.href = "/subscriptions"
         }
      })
   }
}

const addTrialPeriod = (days: number) => {
   try {
      //Logic add trial period to Auth0
      send_welcome_email()
   } catch (e: any) {
      console.log(e.message)
      return e.message
   }
}

const send_welcome_email = () => {
   const body = { "key": process.env.REACT_APP_MAILCHIMP_API_KEY }
   axios({
      headers: {
         'Content-Type': 'application/json',
      },
      method: 'POST',
      url: 'https://mandrillapp.com/api/1.0/users/ping',
      data: body,
   })
      .then(function (response) {
         return {
            code: 'sucess',
            value: response.data,
         }
      })
      .catch(function (error) {
         return {
            code: 'error',
            value: error,
         }
      })
}