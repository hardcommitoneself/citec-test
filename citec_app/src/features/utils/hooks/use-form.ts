// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.

import { useState } from 'react'

interface HandleInputChangeConfig {
   toUpperCase?: boolean
}

export const useForm = <T>(initialFormState: T) => {
   const [formState, setFormState] = useState(initialFormState)

   const handleInputChange = (
      { target }: React.ChangeEvent<HTMLInputElement>,
      config?: HandleInputChangeConfig
   ) => {
      setFormState({
         ...formState,
         [target.name]: config?.toUpperCase
            ? target.value.toUpperCase()
            : target.value,
      })
   }
   const handleInputNumberChange = (name: string, value: string) => {
      setFormState({
         ...formState,
         [name]: value,
      })
   }

   const handleResetForm = () => {
      setFormState(initialFormState)
   }

   return {
      handleInputChange,
      handleInputNumberChange,
      handleResetForm,
      values: formState,
   }
}
