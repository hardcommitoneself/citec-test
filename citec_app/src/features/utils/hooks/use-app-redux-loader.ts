// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import { useDispatch } from 'react-redux'
import { useEffect } from 'react'
import {
   fetchStaticRebalanceData,
   fetchStaticUniverseData,
} from 'store/slices/static-data/thunk'
import { fetchIdeas } from 'store/slices/ideas/thunk'
import { useAppSelector } from 'store/hooks'

export const useAppReduxLoader = (token: string | null) => {
   const dispatch = useDispatch()

   const universes = useAppSelector((state) => state.static_data.universe.universes)
   const ideas = useAppSelector((state) => state.ideas)
   const isUniversesOnRedux = universes.length > 1
   const isIdeasOnRedux = ideas.length > 0

   useEffect(() => {
      //get all static data
      if (token && !isUniversesOnRedux) {
         setTimeout(() => {
            dispatch(fetchStaticUniverseData())
            dispatch(fetchStaticRebalanceData())
         }, 500) // Timeout for Testing app.test.tsx
      }
      if (token && !isIdeasOnRedux) {
         dispatch(fetchIdeas())
      }
   }, [token, universes])

   return null
}
