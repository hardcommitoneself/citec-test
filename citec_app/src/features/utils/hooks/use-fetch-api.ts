// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import { useDispatch } from 'react-redux'
import { useEffect, useState } from 'react'

import { fetchDataFromCitecApi } from '../api/citec-api'
import { setError } from 'store/slices/errors'
import { useAppSelector } from 'store/hooks'

type UseFetchApiProps = {
   endpoint: string
   method: 'POST' | 'GET' | 'DELETE' | 'PUT'
   payload?: any
   params?: any
   startCallInmediatly?: boolean
   dependency?: any
}

export const useFetchApi = <T>({
   endpoint,
   payload,
   method,
   params,
   dependency,
   startCallInmediatly = true,
}: UseFetchApiProps) => {
   const [resp, setResp] = useState<T | null>(null)
   const [isLoading, setIsLoading] = useState(false)
   const error = useAppSelector((state) => state.errors)

   const dispatch = useDispatch()

   useEffect(() => {
      const fetchApi = async () => {
         setIsLoading(true)
         try {
            const { data } = await fetchDataFromCitecApi<T>({
               endpoint,
               method,
               params,
               payload,
            })
            setResp(data)
         } catch (e: any) {
            //timeout error

            if (e.code === 'ECONNABORTED') {
               dispatch(
                  setError(
                     'This process is taking too long, please refresh the page'
                  )
               )
            }

            if (e.response) {
               dispatch(setError(`An error ${e.response.status} has occurred `))
            } else {
               dispatch(
                  setError(
                     'Please go back and verify the configuration selected previously.'
                  )
               )
            }
         }

         setIsLoading(false)
      }

      if (startCallInmediatly) {
         fetchApi()
      }
      return () => {
         setResp(null)
      }
   }, [startCallInmediatly, dependency])

   return { resp, isLoading, error }
}
