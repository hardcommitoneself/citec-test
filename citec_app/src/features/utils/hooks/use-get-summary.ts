// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import { useEffect, useState } from 'react'
import { useDispatch } from 'react-redux'

import { fetchDataFromCitecApi } from '../api/citec-api'
import { setError } from 'store/slices/errors'
import { useAppSelector } from 'store/hooks'
import { BacktestRes } from '../interfaces/backtest-api-resp'
import { PortfolioOptimizationRes } from '../interfaces/portfolio-optimization-res'
import {
   resetPortfolioOptimized,
   setPortfolioOptimized,
} from 'store/slices/portfolio-optimized'
import { resetTrades, setTrades } from 'store/slices/trades'
import { generatePaylaodBacktest } from '../generate-payload-backtest'
import { generatePayloadTrades } from '../generate-payload-trades'
import { generatePayloadOptimization } from '../generate-payload-optimization'
import { PortfolioAnalisysResp } from '../interfaces/portfolio-analysis-resp'
import { Trades } from '../interfaces/trades'
import { setPortfolioAnalysis } from 'store/slices/portfolio'
import { fillTradesGenerated } from 'features/brokers/utils/fill-trades-generate'
import { getPortfolioPositions } from 'features/brokers/utils/get-portfolio-positions'
import { getTickers } from 'features/brokers/alpaca/ui/components/alpaca-link-execute-trades/execute-trades-utils'

type Backtest = Pick<
   BacktestRes,
   'risk' | 'tracking' | 'ts_backtest' | 'tracking_expected'
>

interface SummaryResp {
   backtest: Backtest | null
   analysis: PortfolioAnalisysResp | null
}

export const useGetSummary = () => {
   const [isLoading, setIsLoading] = useState(true)
   const [resps, setResps] = useState<SummaryResp>({
      backtest: null,
      analysis: null,
   })
   const { payloadOptimization, hist_w } = generatePayloadOptimization()
   const error = useAppSelector((state) => state.errors)
   const benchmark = useAppSelector((state) => state.rebalance.benchmark)

   const dispatch = useDispatch()

   const getAllDataSummary = async () => {
      try {

         const optimizationResp = await fetchDataFromCitecApi<PortfolioOptimizationRes>({
               endpoint: '/ai/portfolio/optimization/',
               method: 'POST',
               params: {
                  hist_w,
               },
               payload: payloadOptimization,
         })
         
         const optimization = optimizationResp.data

         dispatch(resetPortfolioOptimized())
         dispatch(setPortfolioOptimized({ ...optimization }))

         const payloadTrades= await generatePayloadTrades()

         const tradesToOrdersPromise = fetchDataFromCitecApi<Trades[]>({
            endpoint: '/ai/portfolio/trades_to_orders',
            method: 'POST',
            params: {
               hist_w: 182,
            },
            payload: payloadTrades,
         })

         const backtestPayload = generatePaylaodBacktest()

         const backtestPromise = fetchDataFromCitecApi<BacktestRes>({
            endpoint: '/ai/portfolio/backtest',
            method: 'POST',
            params: {
               hist_w: 182,
            },
            payload: backtestPayload,
         })

         const analisysPromise = fetchDataFromCitecApi<PortfolioAnalisysResp>({
            endpoint: '/ai/portfolio/analysis',
            method: 'POST',
            params: {
               hist_w: 50000,
            },

            payload: {
               portfolio: 'optimized_portfolio',
               benchmark,
               ticker: optimization.ticker,
               outstanding_balance: optimization.outstanding_balance,
               unrealized_capital_gains: optimization.unrealized_capital_gains,
            },
         })

         const tickers = getTickers(payloadTrades.trades)

         const positionsPromise = getPortfolioPositions(tickers)

         const [backtestResp, analysisResp, tradesToOrdersResp, positionsResp] = await Promise.all([
            backtestPromise,
            analisysPromise,
            tradesToOrdersPromise,
            positionsPromise
         ])

         const { risk, tracking, ts_backtest, tracking_expected } = backtestResp.data
         setResps({
            backtest: {
               risk,
               tracking,
               ts_backtest,
               tracking_expected,
            },
            analysis: analysisResp.data,
         })
         
         const fillTrades = await fillTradesGenerated(tradesToOrdersResp.data, positionsResp, payloadTrades.latest_prices)

         dispatch(setPortfolioAnalysis({ ...analysisResp.data }))
         dispatch(resetTrades())
         dispatch(setTrades(fillTrades))

      } catch (e: any) {
         if (e.code === 'ECONNABORTED') {
            dispatch(
               setError('This process is taking too long, please refresh the page')
            )
         }

         if (e.response) {
            dispatch(setError(`An error ${e.response.status} has occurred `))
         } else {
            dispatch(
               setError(
                  'Please go back and verify the configuration selected previously.'
               )
            )
         }
      }

      setIsLoading(false)
   }

   useEffect(() => {
      getAllDataSummary()
   }, [])

   useEffect(() => {
      if (error) {
         setIsLoading(false)
      }
   }, [error])

   return { isLoading, ...resps }
}
