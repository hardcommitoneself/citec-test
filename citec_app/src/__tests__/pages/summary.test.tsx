// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.

import { BrowserRouter } from 'react-router-dom'
import { exampleReduxState } from 'features/utils/const/exampleReduxState'
import { mount, ReactWrapper } from 'enzyme'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import { useGetSummary } from 'features/utils/hooks/use-get-summary'
import { Summary } from 'pages'

class ResizeObserver {
   observe() {}
   unobserve() {}
}
const mockUseGetSummary = useGetSummary as jest.MockedFunction<typeof useGetSummary>
jest.mock('features/utils/hooks/use-get-summary')

describe('test Summary Page', () => {
   const mockStore = configureStore()
   const store = mockStore(exampleReduxState)

   let wrapper: ReactWrapper<
      any,
      Readonly<unknown>,
      React.Component<unknown, unknown, any>
   >

   beforeEach(async () => {
      mockUseGetSummary.mockReturnValue({
         isLoading: false,
         backtest: {
            risk: {
               sharpe: -1.107185,
               volatility: 0.239029,
               beta: 1.034119,
               alpha: -1e-5,
            },

            tracking: {
               tracking_error: 0.015409,
               tracking_difference: -4.3e-5,
               information_ratio: -0.710898,
               r_square: 0.99693,
               alpha: -1e-5,
               beta: 1.034119,
            },

            ts_backtest: {
               date: ['24-11-2021', '26-11-2021', '29-11-2021'],
               benchmark: [1.00267, 0.980307, 0.992332, 0.973024],
               portfolio: [1.0028, 0.9793, 0.99247],
            },
         },
         analysis: {
            fundamentals: {
               total_assets: 4,
               total_positions: 4,
               pe: 73.739873,
               pb: 28.025954,
               dividend_yield: 0.517303,
            },
            risk: {
               sharpe: -1.002506,
               volatility: 0.068686,
               beta: 0.315348,
               alpha: -0.000072,
            },
            returns: {
               returns_1M: -0.0148,
               returns_3M: -0.0103,
               returns_6M: null,
               CAGR: -0.0575,
            },
            capital_gains: {
               outstanding_balance: 1310000,
               unrealized_cap_losses: 101000,
               unrealized_cap_gains: 11234,
               realized_cap_losses: 0,
               realized_cap_gains: 0,
               taxes_cost: 0,
               taxes_rate: 0.4,
            },
            sectors: {
               sector: ['Cash', 'Consumer Discretionary', 'Information Technology'],
               percentage: [0.763359, 0.007634, 0.229008],
            },
            factors: {
               factor: [
                  'style',
                  'size',
                  'yield',
                  'quality',
                  'volatility',
                  'momentum',
                  'liquidity',
               ],
               percentage: [
                  0.513052, 0.618264, 0.59449, 0.5672, 0.432629, 0.529742, 0.465484,
               ],
            },
            segment_matrix: {
               style: [
                  'Val',
                  'Val',
                  'Val',
                  'Bld',
                  'Bld',
                  'Bld',
                  'Gwth',
                  'Gwth',
                  'Gwth',
               ],
               size: [
                  'Large',
                  'Mid',
                  'Small',
                  'Large',
                  'Mid',
                  'Small',
                  'Large',
                  'Mid',
                  'Small',
               ],
               weight: [0, 0, 0, 19.083969, 0, 0, 4.580153, 0, 0],
            },
            top_holdings: {
               ticker: ['$CASH', 'MSFT', 'AAPL', 'TSLA'],
               weight: [0.763359, 0.19084, 0.038168, 0.007634],
               outstanding_balance: [1000000, 250000, 50000, 10000],
               unrealized_capital_gains: [0, -65000, 11234, -36000],
            },
            concentration: {
               ticker: ['$CASH', 'MSFT', 'AAPL', 'TSLA'],
               concentration: [76.335878, 95.419847, 99.236641, 100],
            },
            var: {
               quantile: [
                  5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75, 80, 85,
                  90, 95,
               ],
               value: [
                  0, 0, 0, 0, 0.010967, 0.021321, 0.030917, 0.040021, 0.04883,
                  0.0575, 0.066169, 0.074978, 0.084083, 0.093678, 0.104032, 0.115563,
                  0.129003, 0.145913, 0.170978,
               ],
            },
            risk_returns: {
               ticker: ['portfolio', 'benchmark'],
               returns: [-0.0575, -0.122369],
               volatility: [0.068686, 0.190293],
               sharpe: [0.837133, 0.643053],
            },
         },
      } as never)

      wrapper = await mount(
         <Provider store={store}>
            <BrowserRouter>
               <Summary />
            </BrowserRouter>
         </Provider>
      )
   })

   test('snapshot', () => {
      window.ResizeObserver = ResizeObserver

      expect(wrapper).toMatchSnapshot()
   })

   test('Initial load is Trade tab', () => {
      window.ResizeObserver = ResizeObserver

      const activeView = wrapper.find({ 'data-testid': 'active-trades' })

      expect(activeView).toBeTruthy()
   })
})
