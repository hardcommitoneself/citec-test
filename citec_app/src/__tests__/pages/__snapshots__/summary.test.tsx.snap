// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Summary Page snapshot 1`] = `ReactWrapper {}`;
